### PR TITLE
CSA-1 - adding master password authentication when enrolling in passw…

### PR DESCRIPTION
**Pull request by cmuentes**
…ord reset

## Type of change
- [ X] Bug fix
- [ ] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other

## Objective
Adding master password authentication when enrolling in password reset

## Code changes

* **OrganizationUserRequestModel.cs:** Added master password hash property
* * **OrganizationUsersController.cs:** Added user password verification

## Testing requirements
Check that the user's master password is required to enroll in password reset

## Before you submit
- [ ] I have checked for formatting errors (`dotnet tool run dotnet-format --check`) (required)
- [ ] If making database changes - I have also updated Entity Framework queries and/or migrations
- [ ] I have added **unit tests** where it makes sense to do so (encouraged but not required)
- [ ] This change requires a **documentation update** (notify the documentation team)
- [ ] This change has particular **deployment requirements** (notify the DevOps team)

### CODE REVIEW

File: src/Api/Controllers/OrganizationUsersController.cs [hunk 1]
```diff
@@ -270,8 +270,17 @@ public async Task PutGroups(string orgId, string id, [FromBody] OrganizationUser
         [HttpPut("{userId}/reset-password-enrollment")]
         public async Task PutResetPasswordEnrollment(string orgId, string userId, [FromBody] OrganizationUserResetPasswordEnrollmentRequestModel model)
         {
+            var user = await _userService.GetUserByIdAsync(new Guid(userId));
```
Comment (kspearrin) [inline src/Api/Controllers/OrganizationUsersController.cs:273]: I think userId and  the `User` principal are the same here (as verified later in `UpdateUserResetPasswordEnrollmentAsync`. In that case, you could optimize this a bit more and just do

```
            var callingUser = await _userService.GetUserByPrincipalAsync(User);
```

and get rid of `_userService.GetProperUserId(User);` below

File: src/Api/Controllers/OrganizationUsersController.cs [hunk 2]
```diff
@@ -270,8 +270,17 @@ public async Task PutGroups(string orgId, string id, [FromBody] OrganizationUser
...
+            var authenticated = await _userService.CheckPasswordAsync(user, model.MasterPasswordHash);
```
Comment (kspearrin) [inline src/Api/Controllers/OrganizationUsersController.cs:274]: And use callingUser here

**Review by kspearrin — CHANGES_REQUESTED** [review]

File: src/Api/Controllers/OrganizationUsersController.cs (hunk 1 above)
Comment (cmuentes) [reply src/Api/Controllers/OrganizationUsersController.cs:273]: Got it - modified to:
`var callingUser = await _userService.GetUserByPrincipalAsync(this.User);
            await _organizationService.UpdateUserResetPasswordEnrollmentAsync(
                new Guid(orgId), new Guid(userId), model.ResetPasswordKey, callingUser.Id);`

File: src/Api/Controllers/OrganizationUsersController.cs (hunk 2 above)
Comment (cmuentes) [reply src/Api/Controllers/OrganizationUsersController.cs:274]: I'm not sure I follow - there appears to be separate users - the user that's enrolling and the user doing the enrolling. It seems like they can be the same by accident, but the intent seems to allow them to vary. Does using the `User` principle bind these together incorrectly or am I misreading something?

Comment (kspearrin) [reply src/Api/Controllers/OrganizationUsersController.cs:274]: In the case of the enrollment, the calling user is always the same as the user being enrolled. Also why we are also asking for their master password, which only they would know.

Comment (cmuentes) [reply src/Api/Controllers/OrganizationUsersController.cs:274]: @kspearrin great question! After reviewing the issue, the only exploitable vulnerability here is if a user walked away from their machine with their vault unlocked, someone could walk by and enroll them in password reset as current the app does not prompt for the master password.

The only time an admin can enroll a user in password reset is when they are invited to an org. After that, the only way to enroll is if the user self-enrolls so I added the reprompt. Does that make sense?

Comment (cmuentes) [reply src/Api/Controllers/OrganizationUsersController.cs:274]: @kspearrin PR has been updated - thanks for the feedback!

File: src/Api/Controllers/OrganizationUsersController.cs [hunk 3]
```diff
@@ -270,8 +270,17 @@ public async Task PutGroups(string orgId, string id, [FromBody] OrganizationUser
...
         {
+            var user = await _userService.GetUserByPrincipalAsync(this.User);
+            var authenticated = await _userService.CheckPasswordAsync(user, model.MasterPasswordHash);
+
+            if (!authenticated)
+            {
+                throw new UnauthorizedAccessException("Invalid user name or password");
+            }
+
             var callingUserId = _userService.GetProperUserId(User);
```
Comment (kspearrin) [inline src/Api/Controllers/OrganizationUsersController.cs:281]: You don't need this anymore since you can just use `user.Id`

**Review by kspearrin — CHANGES_REQUESTED** [review]

**Comment by MGibson1** [comment]
@kspearrin mind taking a look at this in context with bitwarden/jslib#810 and bitwarden/web#1698?

**Review by kspearrin — APPROVED** [review]

**Pull request merged**